src/widgets/modals/StocktakeBatchModal.js

```diff
@@ -1,413 +1,175 @@
-            style={localStyles.reasonCell}
-          >
-            {editable && <Icon name="external-link" size={14} color={SUSSOL_ORANGE} />}
-            <Text style={{ width: '80%' }} numberOfLines={1} ellipsizeMode="tail">
-              {stocktakeBatch.option ? stocktakeBatch.option.title : programStrings.not_applicable}
-            </Text>
-          </TouchableOpacity>
-        );
-      }
+        return PageActions.editStocktakeBatchName;
+      case 'countedTotalQuantity':
+        return PageActions.editStocktakeBatchCountedQuantity;
+      case 'expiryDate':
+        return PageActions.editStocktakeBatchExpiryDate;
+      case 'reasonTitle':
+        return onEditReason;
       default:
-        return {
-          cellContents: stocktakeBatch[key],
-        };
+        return null;
     }
   };
 
-  renderAddBatchButton = () => {
-    const { database, stocktakeItem } = this.props;
-
-    const addNewBatch = () => {
-      database.write(() => {
-        stocktakeItem.createNewBatch(database);
-      });
-      this.refreshData();
-    };
-
-    return (
-      <PageButton
-        text={buttonStrings.add_batch}
-        onPress={addNewBatch}
-        isDisabled={stocktakeItem.stocktake.isFinalised}
-        style={localStyles.addBatchButton}
-      />
-    );
-  };
-
-  renderPageInfo = () => {
-    const { stocktakeItem } = this.props;
-    const { itemName } = stocktakeItem;
-
-    const infoColumns = [
-      [
-        {
-          title: pageInfoStrings.by_batch,
-          info: itemName,
-        },
-      ],
-    ];
-    return <PageInfo columns={infoColumns} />;
-  };
-
-  renderFooter = () => {
-    const { onConfirm } = this.props;
-    return (
-      <View style={localStyles.footer}>
-        <Button
-          text={buttonStrings.done}
-          disabledColor={WARM_GREY}
-          style={[globalStyles.button, localStyles.OKButton]}
-          textStyle={[globalStyles.buttonText, localStyles.OKButtonText]}
-          onPress={onConfirm}
+  const renderRow = useCallback(
+    listItem => {
+      const { item, index } = listItem;
+      const rowKey = keyExtractor(item);
+      return (
+        <DataTableRow
+          rowData={data[index]}
+          rowState={dataState.get(rowKey)}
+          rowKey={rowKey}
+          columns={columns}
+          dispatch={dispatch}
+          getAction={getAction}
+          rowIndex={index}
         />
+      );
+    },
+    [data, dataState]
+  );
+
+  const renderHeader = useCallback(
+    () => (
+      <DataTableHeaderRow
+        columns={columns}
+        dispatch={instantDebouncedDispatch}
+        sortAction={PageActions.sortData}
+        isAscending={isAscending}
+        sortBy={sortBy}
+      />
+    ),
+    [sortBy, isAscending]
+  );
+
+  const PageButtons = () => (
+    <PageButton
+      text={buttonStrings.add_batch}
+      onPress={() => dispatch(PageActions.addStocktakeBatch())}
+      isDisabled={stocktakeItem.stocktake.isFinalised}
```

Comment:
Even though it's in the schema, might want to add a guard here to prevent undefined object error.

Reply:
I think we need to trust our databases constraints, otherwise whats the point? Although I agree defensive code is great, this seems a little far

Reply:
> I think we need to trust our databases constraints, otherwise whats the point? Although I agree defensive code is great, this seems a little far

I'm a big believer in defensive programming in the strict sense (i.e. defend against everything, even if you don't think it could ever happen).

I'd say the point of a schema is to tell the DB what constraints to enforce internally and to give a contract to your code to implement. JS isn't typed, so there's no compile-time constraints on `stocktakeItem`. IMO you trust the DB to enforce it's constraints internally, but you shouldn't assume that also applied to user code, e.g.:


```
  const PageButtons = () => {
    stocktakeItem = undefined;
    return (
      <PageButton
        text={buttonStrings.add_batch}
        onPress={() => dispatch(PageActions.addStocktakeBatch())}
        isDisabled={stocktakeItem.stocktake.isFinalised}
      />
    );
  };
```

Doing this throws an `undefined is not an object` error. Sure, it's a dumb thing to do, but point being that the DB doesn't do anything to prevent this happening.

There's also the fact that there's no guarantee that the schema will remain the same over time, e.g. what if someone renamed `stocktake` to `linkedStocktake` and forgot to update the code?

You could argue that throwing an error is the best thing to do when user code breaks a DB constraint, but for a user, the RN error (red screen of death) is pretty horrific, better to handle it gracefully with defensive code where possible IMO.

Reply:
My point is you can have some guarantees on the schema, but you can't guarantee that a field will be present on a js object, even if it instantiates a realm class. So, I wouldn't suggest defensively coding against everything, just against referencing fields on an object without a guard. I think the whole DB aspect in this instance is a red-herring as it seems like Realm doesn't enforce as much as it maybe should?

Reply:
Made a mistake above, turns out Realm does enforce more than I thought. Realm prevents `Stocktake.stocktakeItem = undefined` (can't write outside a transaction, and even in a transaction it seems to ignore it?), however you can still do `Stocktake = undefined`, which is pretty much a variation on the same deal.

I think we agree it's a balance, I just think making any assumptions about "this won't be a bug" is dangerous and the small investment up-front is worth it in terms of being able to guarantee stable software to the client.